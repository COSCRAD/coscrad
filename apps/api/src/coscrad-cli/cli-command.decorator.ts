/**
 * Given that we have an extensive custom set of decorators for commands in the
 * CQRS sense of the word "command", we re-export and alias `nest-commander` decorators
 * for clarity in naming. These decorators should be imported from this local file
 * into any CLI Command files.
 */
export {
    Command as CliCommand,
    CommandRunner as CliCommandRunner,
    Option as CliCommandOption,
} from 'nest-commander';
