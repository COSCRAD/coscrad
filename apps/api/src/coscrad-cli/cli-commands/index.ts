export * from './performance-test.cli-command';
