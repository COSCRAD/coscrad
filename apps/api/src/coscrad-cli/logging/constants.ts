export const COSCRAD_CLI_LOGGER_TOKEN = `COSCRAD_CLI_LOGGER_TOKEN`;
