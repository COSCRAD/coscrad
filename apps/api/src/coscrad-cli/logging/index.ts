export * from './console-coscrad-cli-logger';
export * from './constants';
export * from './coscrad-cli-logger.interface';
