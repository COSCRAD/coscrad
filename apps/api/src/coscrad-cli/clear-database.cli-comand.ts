import { Inject } from '@nestjs/common';
import { ArangoDatabaseProvider } from '../persistence/database/database.provider';
import { DataImporter } from '../persistence/repositories/data-importer';
import { CliCommand, CliCommandRunner } from './cli-command.decorator';
import { COSCRAD_LOGGER_TOKEN, ICoscradLogger } from './logging';

@CliCommand({
    name: 'clear-database',
    description: '(Cypress mode only) clears database for automated tests',
})
export class ClearDatabaseCliCommand extends CliCommandRunner {
    dataImporter: DataImporter;

    constructor(
        databaseProvider: ArangoDatabaseProvider,
        @Inject(COSCRAD_LOGGER_TOKEN) private readonly logger: ICoscradLogger
    ) {
        super();

        this.dataImporter = new DataImporter(databaseProvider, this.logger);
    }

    // note that this will not work unless in a testing environment with `$DATA_MODE=_CYPRESS`
    async run(_passedParams: string[]) {
        await this.dataImporter.deleteAllData();

        this.logger.log(`Successfully emptied all collections`);
    }
}
