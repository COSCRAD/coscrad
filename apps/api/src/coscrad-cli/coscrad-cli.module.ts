import { CommandModule } from '@coscrad/commands';
import { Module } from '@nestjs/common';
import { AppModule } from '../app/app.module';
import { EdgeConnectionModule } from '../app/domain-modules/edge-connection.module';
import { MediaItemModule } from '../app/domain-modules/media-item.module';
import { VocabularyListModule } from '../app/domain-modules/vocabulary-list.module';
import { EventModule } from '../domain/common';
import { AudioVisualModule } from '../domain/models/audio-visual/application/audio-visual.module';
import { IdGenerationModule } from '../lib/id-generation/id-generation.module';
import { MigrationModule } from '../persistence/migrations';
import { PersistenceModule } from '../persistence/persistence.module';
import { ClearDatabaseCliCommand } from './clear-database.cli-comand';
import { DomainDumpCliCommand } from './data-dump.cli-command';
import { DomainRestoreCliCommand } from './data-restore.cli-command';
import { ExecuteCommandStreamCliCommand } from './execute-command-stream.cli-command';
import { ExportAudioItemLineagesCliCommand } from './export-audio-item-lineages.cli-command';
import { ExportMediaAnnotationsCliCommand } from './export-media-annotations.cli-command';
import { ExportSchemasCliCommand } from './export-schemas.cli-command';
import { IngestMediaItemsCliCommand } from './ingest-media-items.cli-command';
import { ListMigrationsCliCommand } from './list-migrations.cli-command';
import { ConsoleCoscradCliLogger, COSCRAD_LOGGER_TOKEN } from './logging';
import { RehydrateViewsCliCommand } from './rehydrate-views.cli-command';
import { RevertLatestMigrationCliCommand } from './revert-latest-migration';
import { RunMigrationsCliCommand } from './run-migrations.cli-command';
import { SeedDatabaseCliCommand } from './seed-database.cli-command';
import { SeedTestDataWithCommand } from './seed-test-data-with-command.cli-command';
import { SeedTestUuids } from './seed-test-uuids.cli-command';
import { ValidateInvariantsCliCommand } from './validate-invariants.cli-command';

@Module({
    providers: [
        DomainDumpCliCommand,
        DomainRestoreCliCommand,
        ListMigrationsCliCommand,
        RunMigrationsCliCommand,
        RevertLatestMigrationCliCommand,
        ValidateInvariantsCliCommand,
        ClearDatabaseCliCommand,
        ExecuteCommandStreamCliCommand,
        SeedDatabaseCliCommand,
        SeedTestDataWithCommand,
        SeedTestUuids,
        IngestMediaItemsCliCommand,
        ExportMediaAnnotationsCliCommand,
        ExportAudioItemLineagesCliCommand,
        ExportSchemasCliCommand,
        RehydrateViewsCliCommand,
        {
            provide: COSCRAD_LOGGER_TOKEN,
            useClass: ConsoleCoscradCliLogger,
        },
    ],
    imports: [
        AppModule,
        PersistenceModule,
        IdGenerationModule,
        MigrationModule,
        CommandModule,
        VocabularyListModule,
        AudioVisualModule,
        MediaItemModule,
        EdgeConnectionModule,
        EventModule,
    ],
})
export class CoscradCliModule {}
