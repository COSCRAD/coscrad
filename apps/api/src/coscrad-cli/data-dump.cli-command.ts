import { Inject } from '@nestjs/common';
import { writeFileSync } from 'fs';
import { IRepositoryProvider } from '../domain/repositories/interfaces/repository-provider.interface';
import { REPOSITORY_PROVIDER_TOKEN } from '../persistence/constants/persistenceConstants';
import { DomainDataExporter } from '../persistence/repositories/domain-data-exporter';
import convertInMemorySnapshotToDatabaseFormat from '../test-data/utilities/convertInMemorySnapshotToDatabaseFormat';
import { CliCommand, CliCommandOption, CliCommandRunner } from './cli-command.decorator';

@CliCommand({
    name: 'data-dump',
    description: 'dumps the database state to a snapshot file',
})
export class DomainDumpCliCommand extends CliCommandRunner {
    private readonly dataExporter: DomainDataExporter;

    constructor(@Inject(REPOSITORY_PROVIDER_TOKEN) repositoryProvider: IRepositoryProvider) {
        super();

        this.dataExporter = new DomainDataExporter(repositoryProvider);
    }

    async run(_passedParams: string[], options?: { filepath: string }): Promise<void> {
        const NUMBER_OF_SPACES_TO_INDENT = 4;

        const domainSnapshot = await this.dataExporter.fetchSnapshot();

        const snapshotInDatabaseFormat = convertInMemorySnapshotToDatabaseFormat(
            domainSnapshot.fetchFullSnapshotInLegacyFormat()
        );

        writeFileSync(
            options.filepath,
            JSON.stringify(snapshotInDatabaseFormat, null, NUMBER_OF_SPACES_TO_INDENT)
        );
    }

    @CliCommandOption({
        flags: '-f, --filepath [filepath]',
        description: 'the path to write the output to',
        required: true,
    })
    parseFilepath(value: string): string {
        return value;
    }
}
