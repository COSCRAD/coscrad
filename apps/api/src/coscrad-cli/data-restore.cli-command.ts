import { Inject } from '@nestjs/common';
import { readFileSync } from 'fs';
import { InternalError } from '../lib/errors/InternalError';
import { ArangoQueryRunner } from '../persistence/database/arango-query-runner';
import { ArangoDatabaseProvider } from '../persistence/database/database.provider';
import { ArangoDataExporter } from '../persistence/repositories/arango-data-exporter';
import { InMemoryDatabaseSnapshot } from '../test-data/utilities';
import { CliCommand, CliCommandOption, CliCommandRunner } from './cli-command.decorator';
import { COSCRAD_LOGGER_TOKEN, ICoscradLogger } from './logging';

@CliCommand({
    name: 'data-restore',
    description: 'restores the database state from a snapshot file',
})
export class DomainRestoreCliCommand extends CliCommandRunner {
    private readonly dataExporter: ArangoDataExporter;

    constructor(
        databaseProvider: ArangoDatabaseProvider,
        @Inject(COSCRAD_LOGGER_TOKEN) private readonly logger: ICoscradLogger
    ) {
        super();

        this.dataExporter = new ArangoDataExporter(new ArangoQueryRunner(databaseProvider));
    }

    async run(_passedParams: string[], { filepath }: { filepath: string }): Promise<void> {
        const parsedSnapshot = JSON.parse(
            readFileSync(filepath, { encoding: 'utf-8' })
        ) as InMemoryDatabaseSnapshot;

        await this.dataExporter.restoreFromSnapshot(parsedSnapshot).catch((e) => {
            const error = new InternalError(`Data exporter failed to restore data.`, [e]);

            this.logger.log(error.toString());

            throw error;
        });
    }

    @CliCommandOption({
        flags: '-f, --filepath [filepath]',
        description: 'the path to write the output to',
        required: true,
    })
    parseFilepath(value: string): string {
        return value;
    }
}
