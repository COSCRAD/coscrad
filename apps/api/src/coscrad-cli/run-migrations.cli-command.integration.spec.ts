import { TestingModule } from '@nestjs/testing';
import { existsSync, readFileSync, rmSync } from 'fs';
import { CommandTestFactory } from 'nest-commander-testing';
import { AppModule } from '../app/app.module';
import createTestModule from '../app/controllers/__tests__/createTestModule';
import getValidAggregateInstanceForTest from '../domain/__tests__/utilities/getValidAggregateInstanceForTest';
import { CoscradEventFactory } from '../domain/common';
import { MultilingualText } from '../domain/common/entities/multilingual-text';
import buildDummyUuid from '../domain/models/__tests__/utilities/buildDummyUuid';
import { buildFakeTimersConfig } from '../domain/models/__tests__/utilities/buildFakeTimersConfig';
import { AggregateType } from '../domain/types/AggregateType';
import { ResourceType } from '../domain/types/ResourceType';
import { NotFound } from '../lib/types/not-found';
import { REPOSITORY_PROVIDER_TOKEN } from '../persistence/constants/persistenceConstants';
import { ArangoConnectionProvider } from '../persistence/database/arango-connection.provider';
import { ArangoQueryRunner } from '../persistence/database/arango-query-runner';
import { ArangoCollectionId } from '../persistence/database/collection-references/ArangoCollectionId';
import { ArangoDatabaseProvider } from '../persistence/database/database.provider';
import { BASE_DIGITAL_ASSET_URL } from '../persistence/migrations/01/remove-base-digital-asset-url.migration';
import TestRepositoryProvider from '../persistence/repositories/__tests__/TestRepositoryProvider';
import generateDatabaseNameForTestSuite from '../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import { CoscradCliModule } from './coscrad-cli.module';
import { COSCRAD_LOGGER_TOKEN } from './logging';
import { buildMockLogger } from './logging/__tests__';

const cliCommandName = `run-migrations`;

const mockLogger = buildMockLogger();

process.env[BASE_DIGITAL_ASSET_URL] = `https://www.justfortests.io/uploads/`;

const fakeTimersConfig = buildFakeTimersConfig();

/**
 * Note that we have a detailed unit test of the actual
 * `RemoveBaseDigitalAssetUrl` migration. Here, we want to
 * integration test the `run-migrations` cli command. So we
 * are only doing a sanity check that one photograph document
 * gets updated by this command.
 *
 * Further, we do not type the document as we do not want to
 * update this test if we make further changes to the `Photograph`
 * domain model in the future. So we rely on the fact that behind the
 * scenes, the database doesn't know about the domain.
 */
const oldPhotographDocument = {
    _key: buildDummyUuid(1),
    type: ResourceType.photograph,
    filename: `flowers`,
    photographer: `James Rames`,
    dimensions: {
        widthPX: 300,
        heightPX: 400,
    },
    published: true,
};

/**
 * This test has been a source of spontaneous CI failures. We are disabling it for
 * now, as there isn't much risk of regression in this part of the code base.
 *
 * TODO Reinstate this test.
 */
describe.skip(`run migrations`, () => {
    let commandInstance: TestingModule;

    let testRepositoryProvider: TestRepositoryProvider;

    let databaseProvider: ArangoDatabaseProvider;

    beforeAll(async () => {
        const testAppModule = await createTestModule({
            ARANGO_DB_NAME: generateDatabaseNameForTestSuite(),
        });

        await testAppModule.init();

        const arangoConnectionProvider =
            testAppModule.get<ArangoConnectionProvider>(ArangoConnectionProvider);

        databaseProvider = new ArangoDatabaseProvider(arangoConnectionProvider);

        testRepositoryProvider = new TestRepositoryProvider(
            databaseProvider,
            // We don't need the event factory for this test
            new CoscradEventFactory([])
        );

        commandInstance = await CommandTestFactory.createTestingCommand({
            imports: [CoscradCliModule],
        })
            .overrideProvider(AppModule)
            .useValue(testAppModule)
            .overrideProvider(REPOSITORY_PROVIDER_TOKEN)
            .useValue(testRepositoryProvider)
            .overrideProvider(COSCRAD_LOGGER_TOKEN)
            .useValue(mockLogger)
            .overrideProvider(ArangoQueryRunner)
            .useValue(new ArangoQueryRunner(databaseProvider))
            .compile();

        jest.useFakeTimers(fakeTimersConfig);
    });

    // TODO Update the setup and write all other migrations as already run dynamically
    describe(`when there is one migration to run`, () => {
        const dumpDir = `migration-1-RemoveBaseDigitalAssetUrl-${fakeTimersConfig.now}`;

        beforeEach(async () => {
            // clear the database
            await testRepositoryProvider.testSetup();

            // insert a single pre-migration format photograph document for a sanity check
            await databaseProvider
                .getDatabaseForCollection(ArangoCollectionId.photographs)
                .create(oldPhotographDocument);

            // remove data dump dir from previous run
            if (existsSync(dumpDir)) {
                rmSync(dumpDir, { recursive: true, force: true });
            }
        });

        it(`should run the migration with updates to the db`, async () => {
            await CommandTestFactory.run(commandInstance, [cliCommandName]);

            const updatedDocument = await databaseProvider
                .getDatabaseForCollection(ArangoCollectionId.photographs)
                .fetchById(oldPhotographDocument._key);

            expect((updatedDocument as unknown as { imageUrl: string }).imageUrl).toBe(
                `https://www.justfortests.io/uploads/flowers.png`
            );

            expect((updatedDocument as unknown as { filename: string }).filename).toBeUndefined();

            const migrationSequenceNumber = 1;

            const migrationRecord = await databaseProvider
                .getDatabaseForCollection(ArangoCollectionId.migrations)
                .fetchById(migrationSequenceNumber.toString());

            expect(migrationRecord).not.toBe(NotFound);

            delete migrationRecord['_rev'];

            expect(migrationRecord).toMatchSnapshot();
        });

        it(`should write a pre data-dump file`, async () => {
            await CommandTestFactory.run(commandInstance, [cliCommandName]);

            const preFilepath = `${dumpDir}/pre.data.json`;

            expect(existsSync(preFilepath)).toBe(true);

            /**
             * We manually checked the snapshot once. There's not a lot of value
             * in manually testing the contents of the file at the level of an
             * integration test.
             *
             * TODO [test-coverage] [https://github.com/COSCRAD/coscrad/pull/381#discussion_r1198013712]
             * We should do so in a unit test of `ArangoDataExporter`
             */
        });

        it(`should write a post data-dump file`, async () => {
            await CommandTestFactory.run(commandInstance, [cliCommandName]);

            const postFilepath = `${dumpDir}/post.data.json`;

            expect(existsSync(postFilepath)).toBe(true);

            /**
             * We manually checked the snapshot once. There's not a lot of value
             * in manually testing the contents of the file at the level of an
             * integration test. We should do so in a unit test of `ArangoDataExporter`
             */
        });

        describe(`the verification report`, () => {
            describe(`when the resulting database state is valid`, () => {
                it(`should write a verification report`, async () => {
                    await CommandTestFactory.run(commandInstance, [cliCommandName]);

                    const verificationFilePath = `${dumpDir}/verification.data.json`;

                    expect(existsSync(verificationFilePath)).toBe(true);
                });
            });

            describe(`when there are invariant validation errors`, () => {
                beforeEach(async () => {
                    // add an invalid term to the db
                    await testRepositoryProvider.addResourcesOfSingleType(ResourceType.term, [
                        getValidAggregateInstanceForTest(AggregateType.term).clone({
                            // no text in any language
                            text: new MultilingualText({
                                items: [],
                            }),
                        }),
                    ]);
                });

                it(`should write the errors`, async () => {
                    await CommandTestFactory.run(commandInstance, [cliCommandName]);

                    const verificationFilePath = `${dumpDir}/verification.data.json`;

                    const readResult = JSON.parse(
                        readFileSync(verificationFilePath, { encoding: 'utf-8' })
                    );

                    expect(readResult['status']).toBe('failure');
                });
            });
        });
    });
});
