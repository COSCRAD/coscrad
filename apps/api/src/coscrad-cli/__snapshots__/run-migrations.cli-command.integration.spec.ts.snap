// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run migrations when there is one migration to run should run the migration with updates to the db 1`] = `
{
  "_id": "migrations/1",
  "_key": "1",
  "_rev": "_gAshz4i---",
  "dateApplied": 1684359337893,
  "metadata": {
    "dateAuthored": "20230513",
    "description": "convert legacy term and photograph media urls from relative to absolute paths and append extensions",
  },
  "name": "RemoveBaseDigitalAssetUrl",
  "sequenceNumber": 1,
}
`;
